javascript/ql/src/semmle/javascript/security/performance/SuperlinearBackTracking.qll

```diff
@@ -0,0 +1,141 @@
+/**
+ * Provides classes for working with regular expressions that can
+ * perform backtracking in superlinear time.
+ */
+
+import javascript
+
+/**
+ * A regular expression term that permits unlimited repetitions.
+ */
+private class InfiniteRepetitionQuantifier extends RegExpQuantifier {
+  InfiniteRepetitionQuantifier() {
+    this instanceof RegExpPlus
+    or
+    this instanceof RegExpStar
+    or
+    this instanceof RegExpRange and not exists(this.(RegExpRange).getUpperBound())
+  }
+}
+
+/**
+ * Holds if `t` matches at least an epsilon symbol.
```

Comment:
I think this is already available through `RegExpTerm.isNullable`

Reply:
Almost, but not exactly. The two `^$` anchors in particular are nullable, but they do not match "at least an epsilon symbol", they match the implicit start and end symbols instead.

Reply:
> this term does not restrict the language of the enclosing regular expression

I think this makes it clear that `^` and `$` are not epsilon matchers.